registry/storage/driver/hdfsweb/hdfs.go

```diff
@@ -0,0 +1,370 @@
+// store blobs in Hadoop HDFS cloud storage.
+//
+// This package leverages the vladimirvivien/gowfs client functions for
+// interfacing with Hadoop HDFS.
+package hdfsweb
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"reflect"
+	"time"
+
+	"github.com/docker/distribution/context"
+	storagedriver "github.com/docker/distribution/registry/storage/driver"
+	"github.com/docker/distribution/registry/storage/driver/base"
+	"github.com/docker/distribution/registry/storage/driver/factory"
+)
+
+const (
+	driverName         = "hdfsweb"
+	defaultBufferSize  = 4096
+	defaultBlockSize   = 64 << 20
+	defaultReplication = 1
+)
+
+//DriverParameters A struct that encapsulates all of the driver parameters after all values have been set
+type DriverParameters struct {
+	NameNodeHost  string
+	NameNodePort  string
+	RootDirectory string
+	BlockSize     int
+	BufferSize    int
+	Replication   int
+}
+
+func init() {
+	factory.Register(driverName, &hdfsDriverFactory{})
+}
+
+// hdfsDriverFactory implements the factory.StorageDriverFactory interface
+type hdfsDriverFactory struct{}
+
+func (factory *hdfsDriverFactory) Create(parameters map[string]interface{}) (storagedriver.StorageDriver, error) {
+	return FromParameters(parameters)
+}
+
+type driver struct {
+	fs             *FileSystem
+	RootDirectory  string
+	LocalDirectory string
+}
+
+type baseEmbed struct {
+	base.Base
+}
+
+// Driver is a storagedriver.StorageDriver implementation backed by a local
+// hdfs. All provided paths will be subpaths of the RootDirectory.
+type Driver struct {
+	baseEmbed
+}
+
+// FromParameters constructs a new Driver with a given parameters map
+// Optional Parameters:
+func FromParameters(parameters map[string]interface{}) (*Driver, error) {
+	params := DriverParameters{
+		BlockSize:   defaultBlockSize,
+		BufferSize:  defaultBufferSize,
+		Replication: defaultReplication,
+	}
+
+	if parameters != nil {
+		if v, ok := parameters["namenodehost"]; ok {
+			params.NameNodeHost = fmt.Sprint(v)
+		} else {
+			return nil, fmt.Errorf("parameter doesn't provided")
+		}
+
+		if v, ok := parameters["namenodeport"]; ok {
+			params.NameNodePort = fmt.Sprint(v)
+		} else {
+			return nil, fmt.Errorf("parameter doesn't provided")
+		}
+
+		if v, ok := parameters["rootdirectory"]; ok {
+			params.RootDirectory = fmt.Sprint(v)
+		} else {
+			return nil, fmt.Errorf("parameter doesn't provided")
+		}
+
+		if v, ok := parameters["blocksize"]; ok {
+			params.BlockSize = v.(int)
+		}
+
+		if v, ok := parameters["buffersize"]; ok {
+			params.BufferSize = v.(int)
+		}
+
+		if v, ok := parameters["replication"]; ok {
+			params.Replication = v.(int)
+		}
+
+	}
+	return New(params)
+}
+
+// New constructs a new Driver with given parameters
+func New(params DriverParameters) (*Driver, error) {
+	usr, err := user.Current()
+	if err != nil {
+		return nil, err
+	}
+
+	tempDir := filepath.Join(os.TempDir(), "v2")
```

Comment:
I think this should use `ioutil.TempDir` instead. Using a fixed path under /tmp like this has security risks, because an attacker can do things like `ln -s /etc /tmp/v2`.